File: gnss_navigation/gnss_navigation/gnss_path_publisher.py
import rclpy
from rclpy.node import Node
from nav_msgs.msg import Path
from geometry_msgs.msg import PoseStamped
from pyproj import Proj, transform
import pandas as pd
from scipy.interpolate import splprep, splev
import numpy as np

class gnss_path_Publisher(Node):
    def __init__(self):
        super().__init__('gnss_path_Publisher')

        # WGS84座標系（GPSの座標系）
        self.wgs84 = Proj(init='epsg:4326')

        # UTM座標系への変換
        self.utm = Proj(init='epsg:32654')  # UTMゾーンに応じてEPSGコードを変更

        # parameterの設定 
        self.declare_parameter('file_path', '/home/ros2_ws/src/AIFormula_private/gnss_navigation/config/course_data/gazebo_shihou_course.csv')
        file_path = self.get_parameter('file_path').get_parameter_value().string_value
       
        df = pd.read_csv(file_path, header=None)     
        
        self.path = Path()
        self.smooth_path_ = Path()
        self.path.header.stamp = self.get_clock().now().to_msg()
        self.path.header.frame_id = 'map'
        
        for index, data in df.iterrows():
            # 緯度経度をロボット座標系に変換
            x, y = self.convert_gps_to_utm(data[1], data[0])

            if index == 0:
                base_x = x
                base_y = y
            
            # PoseStampedを作成し，経路に追加
            pose = PoseStamped()
            pose.header.stamp = self.get_clock().now().to_msg()
            pose.header.frame_id = 'map'
            pose.pose.position.x = x - base_x
            pose.pose.position.y = y - base_y
            self.path.poses.append(pose)

        self.get_logger().info("data loaded")
        self.get_logger().info("publish start")
            
        # 経路の平滑化処理
        self.smooth_path()

        self.publisher_ = self.create_publisher(Path, 'gnss_path', 10)
        self.timer = self.create_timer(1, self.publish_path)
        
    def convert_gps_to_utm(self, lat, lon):
        x, y = transform(self.wgs84, self.utm, lat, lon)
        return x, y
        
    def smooth_path(self):
        x = [pose.pose.position.x for pose in self.path.poses]
        y = [pose.pose.position.y for pose in self.path.poses]

        # スプライン補間
        tck, u = splprep([x, y], s=0, per=0)
        u_new = np.linspace(u.min(), u.max(), 500)
        x_new, y_new = splev(u_new, tck, der=0)

        # 補間した座標から新しいパスを生成
        self.smooth_path_.header = self.path.header
        for x, y in zip(x_new, y_new):
            pose = PoseStamped()
            pose.pose.position.x = x
            pose.pose.position.y = y
            self.smooth_path_.poses.append(pose)
        
    def publish_path(self):
        self.publisher_.publish(self.smooth_path_)
        
def main(args=None):
    try:
        rclpy.init(args=args)
        pub = gnss_path_Publisher()
        rclpy.spin(pub)
    except KeyboardInterrupt:
        print("ctrl-C")
    finally:
        pub.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: gnss_navigation/gnss_navigation/gnss_path_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "nav_msgs/msg/path.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include <Eigen/Dense>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <proj.h>
#include <unsupported/Eigen/Splines> 

class GNSSPathPublisher : public rclcpp::Node {
public:
    GNSSPathPublisher() : Node("gnss_path_publisher") {
        // パラメータの取得
        this->declare_parameter<std::string>("file_path", "/home/ros2_ws/src/AIFormula_private/gnss_navigation/config/course_data/gazebo_shihou_course.csv");
        auto file_path = this->get_parameter("file_path").as_string();
        // CSVファイルの読み込み
        std::ifstream file(file_path);
        std::string line;
        std::vector<Eigen::Vector2d> coordinates;
        std::vector<Eigen::Vector2d> origin_coordinates;
        std::vector<double> xs;
        std::vector<double> ys;
        std::vector<double> origin_xs;
        std::vector<double> origin_ys;
        
        int cnt=0;
        double base_x;
        double base_y;
        while (std::getline(file, line)) {
            std::stringstream ss(line);
            std::string cell;
            std::vector<std::string> tokens;
            while (std::getline(ss, cell, ',')) {
                tokens.push_back(cell);
            }
            double lat = std::stod(tokens[0]);
            double lon = std::stod(tokens[1]);
            // GPS座標をUTM座標に変換
            auto [x, y] = convertGPStoUTM(lat, lon);
            if (cnt == 0){
                base_x = x;
                base_y = y;
            }

            origin_coordinates.emplace_back(x, y);
            origin_xs.push_back(x);
            origin_ys.push_back(y);

            coordinates.emplace_back(x - base_x, y - base_y);
            xs.push_back(x - base_x);
            ys.push_back(y - base_y);
            cnt++;
   
        }
        

        std::vector<Eigen::Vector2d> spline_points = interpolateSpline(xs, ys, 100);  // 100は補間点の数
        std::vector<Eigen::Vector2d> origin_spline_points = interpolateSpline(origin_xs, origin_ys, 100);  // 100は補間点の数

        // Pathメッセージの初期化
        
        path_msg.header.stamp = this->now();
        path_msg.header.frame_id = "map";
        for (const auto& coord : spline_points) {
            geometry_msgs::msg::PoseStamped pose;
            pose.header.stamp = this->now();
            pose.header.frame_id = "map";
            pose.pose.position.x = coord.x();
            pose.pose.position.y = coord.y();
            path_msg.poses.push_back(pose);
        }
        publisher_ = this->create_publisher<nav_msgs::msg::Path>("gnss_path", 10);

        origin_path_msg.header.stamp = this->now();
        origin_path_msg.header.frame_id = "map";
        for (const auto& origin_coord : origin_spline_points) {
            geometry_msgs::msg::PoseStamped pose;
            pose.header.stamp = this->now();
            pose.header.frame_id = "map";
            pose.pose.position.x = origin_coord.x();
            pose.pose.position.y = origin_coord.y();
            origin_path_msg.poses.push_back(pose);
        }
        origin_publisher_ = this->create_publisher<nav_msgs::msg::Path>("origin_gnss_path", 10);

        _pub_timer = this->create_wall_timer(
            std::chrono::seconds(1),
            [this] { _publisher_callback(); }
        );
    }
    void _publisher_callback() {
            publisher_->publish(path_msg);
            origin_publisher_->publish(origin_path_msg);
    }
private:
    rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr publisher_;
    rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr origin_publisher_;
    rclcpp::TimerBase::SharedPtr _pub_timer;
    nav_msgs::msg::Path path_msg;
    nav_msgs::msg::Path origin_path_msg;
    std::pair<double, double> convertGPStoUTM(double lat, double lon) {
        if (!(-90 <= lat) || !(lat <= 90) || !(-180 <= lon) || !(lon <= 180)) {
        std::cerr << "Error: Latitude or longitude values are out of valid range." << std::endl;
        return {std::numeric_limits<double>::infinity(), std::numeric_limits<double>::infinity()};
    }
    PJ *P = proj_create_crs_to_crs(PJ_DEFAULT_CTX, "EPSG:4326", "EPSG:32654", nullptr);
    PJ_COORD p = proj_coord(lat, lon, 0, 0);
    p = proj_trans(P, PJ_FWD, p);
    proj_destroy(P);
    return {p.xy.x, p.xy.y};
    }

    std::vector<Eigen::Vector2d> interpolateSpline(const std::vector<double>& xs, const std::vector<double>& ys, int num_points) {
    Eigen::Matrix<double, Eigen::Dynamic, 2> points(xs.size(), 2);
    for (size_t i = 0; i < xs.size(); ++i) {
        points(i, 0) = xs[i];
        points(i, 1) = ys[i];
    }

    auto spline = Eigen::SplineFitting<Eigen::Spline<double, 2>>::Interpolate(points.transpose(), 2); // 2次のキュービックスプライン

    std::vector<Eigen::Vector2d> result;
    double step = 1.0 / (num_points - 1);
    for (int i = 0; i < num_points; ++i) {
        double u = i * step;
        Eigen::Vector2d pt = spline(u);
        result.push_back(pt);
    }
    return result;
    }

};
int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<GNSSPathPublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
